mm2src/crypto/src/hw_client.rs

```diff
@@ -139,11 +139,33 @@ impl HwClient {
             Ok(Some(trezor))
         }
 
+        #[cfg(feature = "trezor-udp")]
+        fn try_to_connect_emulator() -> HwResult<Option<TrezorClient>> {
+            let mut devices = trezor::transport::udp::find_devices()?;
+            if devices.is_empty() {
+                return Ok(None);
+            }
+            if devices.len() != 1 {
+                return MmError::err(HwError::CannotChooseDevice { count: devices.len() });
+            }
+            let device = devices.remove(0);
+            let transport = device.connect()?;
+            let trezor = TrezorClient::from_transport(transport);
+            Ok(Some(trezor))
+        }
```

Comment:
Yes I thought about this. It would require to refactor some trezor code and add new traits and generics.
As I initially considered the udp transport as not part of the common test system but rather an optional personal tool for debugging trezor code I decided not to do this.
But as we spoke about adding the trezor emulator to the CI we probably should do this refactoring.

Reply:
added ConnectableDeviceWrapper trait

Reply:
Please consider the below changes related to `ConnectableDeviceWrapper`:
1. The connect method signature now borrows `self` instead of consuming `self`. This required to derive `Clone` and `Copy` for `UsbDeviceInfo` and `UsbDeviceInterfaceInfo` which is acceptable since they contain primitive types only.
2. `T` was renamed to `TransportType` for better readability. It's always better to give trait associated types a clear name unlike function/structs generics where using `T` is acceptable. Trait associated types always have a specific role that's why it's prefered to name them.
3. Since `ConnectableDeviceWrapper` is async, constraining `TransportType` to be send and sync in trait definition is better. Also, it's good to constraint it to implement `Transport`.

```diff
Subject: [PATCH] wip
---
Index: mm2src/trezor/src/transport/udp.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/mm2src/trezor/src/transport/udp.rs b/mm2src/trezor/src/transport/udp.rs
--- a/mm2src/trezor/src/transport/udp.rs	(revision 0f45e36de10dc2bb219484e0daadda7c8e8e24f1)
+++ b/mm2src/trezor/src/transport/udp.rs	(date 1697626282629)
@@ -43,8 +43,8 @@
 
 impl UdpAvailableDevice {
     /// Connect to the device.
-    async fn connect(self) -> TrezorResult<UdpTransport> {
-        let transport = UdpTransport::connect(&self).await?;
+    async fn connect(&self) -> TrezorResult<UdpTransport> {
+        let transport = UdpTransport::connect(self).await?;
         Ok(transport)
     }
 }
@@ -160,14 +160,14 @@
 
 #[async_trait]
 impl ConnectableDeviceWrapper for UdpAvailableDevice {
-    type T = UdpTransport;
+    type TransportType = UdpTransport;
 
     async fn find_devices() -> TrezorResult<Vec<Self>>
     where
         Self: Sized,
     {
-        crate::transport::udp::find_devices().await
+        find_devices().await
     }
 
-    async fn connect(self) -> TrezorResult<UdpTransport> { UdpAvailableDevice::connect(self).await }
+    async fn connect(&self) -> TrezorResult<Self::TransportType> { UdpAvailableDevice::connect(self).await }
 }
Index: mm2src/trezor/src/transport/mod.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/mm2src/trezor/src/transport/mod.rs b/mm2src/trezor/src/transport/mod.rs
--- a/mm2src/trezor/src/transport/mod.rs	(revision 0f45e36de10dc2bb219484e0daadda7c8e8e24f1)
+++ b/mm2src/trezor/src/transport/mod.rs	(date 1697626282626)
@@ -72,13 +72,11 @@
 /// Wrapper to abstract connectivity to usb and emulator devices
 #[async_trait]
 pub trait ConnectableDeviceWrapper {
-    type T;
+    type TransportType: Transport + Sync + Send;
 
     async fn find_devices() -> TrezorResult<Vec<Self>>
     where
         Self: Sized;
 
-    async fn connect(self) -> TrezorResult<Self::T>
-    where
-        Self::T: Transport;
+    async fn connect(&self) -> TrezorResult<Self::TransportType>;
 }
Index: mm2src/trezor/src/transport/usb.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/mm2src/trezor/src/transport/usb.rs b/mm2src/trezor/src/transport/usb.rs
--- a/mm2src/trezor/src/transport/usb.rs	(revision 0f45e36de10dc2bb219484e0daadda7c8e8e24f1)
+++ b/mm2src/trezor/src/transport/usb.rs	(date 1697626718504)
@@ -73,7 +73,7 @@
 
 impl UsbAvailableDevice {
     /// Please note [`hw_common::transport::libusb::UsbAvailableDevice::connect`] spawns a thread.
-    async fn connect(self) -> TrezorResult<UsbTransport> {
+    async fn connect(&self) -> TrezorResult<UsbTransport> {
         let link = UsbLink {
             device: self.0.connect()?,
         };
@@ -95,14 +95,14 @@
 
 #[async_trait]
 impl ConnectableDeviceWrapper for UsbAvailableDevice {
-    type T = UsbTransport;
+    type TransportType = UsbTransport;
 
     async fn find_devices() -> TrezorResult<Vec<Self>>
     where
         Self: Sized,
     {
-        crate::transport::usb::find_devices().await
+        find_devices().await
     }
 
-    async fn connect(self) -> TrezorResult<UsbTransport> { UsbAvailableDevice::connect(self).await }
+    async fn connect(&self) -> TrezorResult<Self::TransportType> { UsbAvailableDevice::connect(self).await }
 }
Index: mm2src/crypto/src/hw_client.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/mm2src/crypto/src/hw_client.rs b/mm2src/crypto/src/hw_client.rs
--- a/mm2src/crypto/src/hw_client.rs	(revision 0f45e36de10dc2bb219484e0daadda7c8e8e24f1)
+++ b/mm2src/crypto/src/hw_client.rs	(date 1697626282633)
@@ -124,12 +124,11 @@
     ) -> MmResult<TrezorClient, HwProcessingError<Processor::Error>> {
         use common::custom_futures::timeout::TimeoutError;
         use common::executor::Timer;
-        use trezor::transport::{ConnectableDeviceWrapper, Transport};
+        use trezor::transport::ConnectableDeviceWrapper;
 
         async fn try_to_connect<C>() -> HwResult<Option<TrezorClient>>
         where
-            C: ConnectableDeviceWrapper,
-            <C as ConnectableDeviceWrapper>::T: Transport + Sync + Send + 'static,
+            C: ConnectableDeviceWrapper + 'static,
         {
             let mut devices = C::find_devices().await?;
             if devices.is_empty() {
Index: mm2src/hw_common/src/transport/libusb.rs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/mm2src/hw_common/src/transport/libusb.rs b/mm2src/hw_common/src/transport/libusb.rs
--- a/mm2src/hw_common/src/transport/libusb.rs	(revision 0f45e36de10dc2bb219484e0daadda7c8e8e24f1)
+++ b/mm2src/hw_common/src/transport/libusb.rs	(date 1697626620395)
@@ -127,7 +127,7 @@
 }
 
 impl UsbAvailableDevice {
-    pub fn connect(self) -> UsbResult<UsbDevice> {
+    pub fn connect(&self) -> UsbResult<UsbDevice> {
         // This is a non-blocking function; no requests are sent over the bus.
         let mut device_handle = self.device.open().map_to_mm(UsbError::ErrorOpeningDevice)?;
         // Claiming of interfaces is a purely logical operation.
@@ -277,7 +277,7 @@
     fn endpoint_number(&self) -> u8 { self.device_info.interface_info.endpoint_number }
 }
 
-#[derive(Debug)]
+#[derive(Clone, Copy, Debug)]
 pub struct UsbDeviceInfo {
     pub vendor_id: u16,
     pub product_id: u16,
@@ -286,7 +286,7 @@
     pub interface_info: UsbDeviceInterfaceInfo,
 }
 
-#[derive(Debug)]
+#[derive(Clone, Copy, Debug)]
 pub struct UsbDeviceInterfaceInfo {
     pub interface_number: u8,
     pub endpoint_number: u8,
```

Reply:
added as it is